### PR TITLE
Fix #172- Re-branding / white labeling app

### LINKED ISSUE #172: Support re-branding/white-labeling without forking project
This has come up multiple times - people want to deploy OBA Android for their own transit agency, but want to re-brand it with colors/logos and even custom icons and Google Play store listings.  It would be nice to be able to support re-branding within the application, without forking the project.  This way everyone continues to benefit from improvements to the app.

I see two possible solutions:

1.  Dynamic re-skinning at run time inside the app

2.  Use Gradle build flavors to manage separate APK creation for each brand

Option 1 is the easiest to manage - you’d still have a single “OneBusAway” app on the market.  When the user downloaded the app, the presentation to the user would be re-skinned with whatever colors and name should be shown to the user (similar to how a region is selected at runtime).  However, this sacrifices the individual brand presence on the app stores.  In Option 2, you’d have individual brand presence on the markets, but at the cost of pushing out a new app version per agency per release.  Pushing out new branded version would need to be handled by whoever controlled that brand, and each brand would need their own Google Play developer account.  Including build flavors in the project also slightly increases complexity of the project.

So, whether or not an agency wants their brand to extend to a separate app on Google Play that users download (i.e., “Bear Transit”, with the Bear Transit logo, instead of OBA name/logo) would likely determine whether Option 1 is acceptable.  From existing conversations, it seems like agencies want Option 2.

For Option 2, see these links for discussions of using build flavors to white-label apps:

* http://apporia.blogspot.com/2013/11/gradle-and-flavor-groups.html

* http://stackoverflow.com/questions/20976946/dynamically-generating-product-flavors

* https://plus.google.com/+CyrilMottier/posts/Dk65EACXqsh

* http://blog.robustastudio.com/mobile-development/android/building-multiple-editions-of-android-app-gradle/

See discussion at:
https://groups.google.com/d/msg/onebusaway-developers/OWkt-1TjRzo/G-fUpXBB-UEJ

### PR DESCRIPTION
**Pull request by barbeau**
This is a working version of Option 2 (as described in #172), rebranding using Gradle build flavor variants, to allow for separately published apps on Google Play (as well as Amazon App Store, etc).

It has two sample agencies, "Agency X" and "Agency Y".  Here are what the rebranding samples look like (with current OBA `develop` branch version to the far left):

brands

You can see via the below screenshot that all three apps can coexist on the same device:

homescreen-small

There is still some small polishing TODO, but it's mostly implemented.  

TODO:
- [x] Allow variants to be permanently fixed to a single API endpoint, instead of searching for regions
- [x] Related to a fixed endpoint - hide some of the Settings related to regions, etc. for variants
- [x] Add customization of arrival times as an example of how a region could change presentation of data
- [x] Fix bug that breaks amazon and google flavors from the command line - right now it seems to compile/run fine in Android Studio, but fails on a BuildConfig.DATABASES_AUTHORITY compile error when doing `gradlew` from the command line, since this variable isn't being read correctly from the build flavors
- [x] Better shading for blue/red stop markers (accent color should be darker)
- [x] Modify Travis build script - right now Travis is trying to execute the unit tests for all 6 build variants (3 brands x 2 platforms), which takes far too long and times out.  Since there aren't (currently) any code differences between the brands, we can change Travis to simply run the 2 platforms for the generic OBA brand (as it was before), which will reduce the total Travis execution time back down to a reasonable number.
- [x] Add some acknowledgement to OBA in the branded variants - something like "Powered by OneBusAway"?
- [x] Change default "on-time" color for Agency X (currently its the primary theme color, which is red, and close to the "early" color)
- [x] Fix a crash on the Agency Y flavor when tapping on arrival time entry
- [x] Fix Agency Y arrival info/map button taps to show tapped state
- [x] Fix stop shortcuts to home screen for Agency Y variant
- [x] Fix the arrival time info button on the Agency Y variant
- [x] Fix a crash on the Agency Y flavor that occurs occasionally when tapping on multiple stops in a row to view arrival times
- [x] Tweak Style B to look like Cards
- [x] If no estimate, show blank on right column for Style B
- [x] Documentation in markdown format

**Please don't merge yet**

### CODE REVIEW

**Comment by barbeau** [comment]
Crash that I'm seeing in Agency Y (arrival info style B) arrival list:

~~~
05-14 19:36:51.142  32360-32747/org.agencyy.android D/ObaContext﹕ Using region base URL 'http://api.tampa.onebusaway.org/api'.
05-14 19:36:51.142  32360-32747/org.agencyy.android D/ObaDefaultConnection﹕ http://api.tampa.onebusaway.org/api/api/where/arrivals-and-departures-for-stop/Hillsborough%20Area%20Regional%20Transit_6649.json?minutesAfter=35&app_ver=38&app_uid=176a7ab5c0fd5ebf83bb7853e8207bf0&version=2&key=v1_BktoDJ2gJlu6nLM6LsT9H8IUbWc%3DcGF1bGN3YXR0c0BnbWFpbC5jb20%3D
05-14 19:36:51.174  32360-32360/org.agencyy.android D/BubblePopupHelper﹕ isShowingBubblePopup : false
05-14 19:36:51.174  32360-32360/org.agencyy.android D/BubblePopupHelper﹕ isShowingBubblePopup : false
05-14 19:36:51.471  32360-32360/org.agencyy.android D/AndroidRuntime﹕ Shutting down VM
05-14 19:36:51.499  32360-32360/org.agencyy.android E/AndroidRuntime﹕ FATAL EXCEPTION: main
    Process: org.agencyy.android, PID: 32360
    java.lang.IndexOutOfBoundsException: Invalid index 0, size is 0
            at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
            at java.util.ArrayList.get(ArrayList.java:308)
            at org.onebusaway.android.ui.ArrivalsListAdapterStyleB.initView(ArrivalsListAdapterStyleB.java:99)
            at org.onebusaway.android.ui.ArrivalsListAdapterStyleB.initView(ArrivalsListAdapterStyleB.java:39)
            at org.onebusaway.android.util.ArrayAdapter.getView(ArrayAdapter.java:70)
            at android.widget.HeaderViewListAdapter.getView(HeaderViewListAdapter.java:220)
            at android.widget.AbsListView.obtainView(AbsListView.java:2387)
            at android.widget.ListView.measureHeightOfChildren(ListView.java:1270)
            at android.widget.ListView.onMeasure(ListView.java:1182)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5464)
            at android.widget.FrameLayout.onMeasure(FrameLayout.java:430)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5464)
            at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1438)
            at android.widget.LinearLayout.measureVertical(LinearLayout.java:724)
            at android.widget.LinearLayout.onMeasure(LinearLayout.java:615)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5464)
            at android.widget.FrameLayout.onMeasure(FrameLayout.java:430)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5464)
            at android.widget.LinearLayout.measureChildBeforeLayout(LinearLayout.java:1438)
            at android.widget.LinearLayout.measureHorizontal(LinearLayout.java:1085)
            at android.widget.LinearLayout.onMeasure(LinearLayout.java:617)
            at android.view.View.measure(View.java:17496)
            at android.widget.LinearLayout.measureVertical(LinearLayout.java:877)
            at android.widget.LinearLayout.onMeasure(LinearLayout.java:615)
            at android.view.View.measure(View.java:17496)
            at com.sothree.slidinguppanel.SlidingUpPanelLayout.onMeasure(SlidingUpPanelLayout.java:649)
            at android.view.View.measure(View.java:17496)
            at android.widget.RelativeLayout.measureChildHorizontal(RelativeLayout.java:728)
            at android.widget.RelativeLayout.onMeasure(RelativeLayout.java:464)
            at android.view.View.measure(View.java:17496)
            at android.support.v4.widget.DrawerLayout.onMeasure(DrawerLayout.java:851)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5464)
            at android.widget.FrameLayout.onMeasure(FrameLayout.java:430)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5464)
            at android.support.v7.internal.widget.ActionBarOverlayLayout.onMeasure(ActionBarOverlayLayout.java:453)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5464)
            at android.widget.FrameLayout.onMeasure(FrameLayout.java:430)
            at android.view.View.measure(View.java:17496)
            at android.widget.LinearLayout.measureVertical(LinearLayout.java:877)
            at android.widget.LinearLayout.onMeasure(LinearLayout.java:615)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewGroup.measureChildWithMargins(ViewGroup.java:5464)
            at android.widget.FrameLayout.onMeasure(FrameLayout.java:430)
            at com.android.internal.policy.impl.PhoneWindow$DecorView.onMeasure(PhoneWindow.java:2636)
            at android.view.View.measure(View.java:17496)
            at android.view.ViewRootImpl.performMeasure(ViewRootImpl.java:2031)
            at android.view.ViewRootImpl.measureHierarchy(ViewRootImpl.java:1193)
            at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:1400)
            at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1078)
            at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:5875)
            at android.view.Choreographer$CallbackRecord.run(Choreographer.java:767)
            at android.view.Choreographer.doCallbacks(Choreographe
~~~

**Comment by barbeau** [comment]
This crash seems to be happening when there is only one arrival time for a given stop, and that arrival time is negative (i.e., the bus just left).  Style B does not show these negative times and filters them out in the adapter, and the adapter ends up with an empty list that it tries to access.

**Comment by barbeau** [comment]
For reference, the arrival style B is an alternate presentation of arrival info that was created by York Region Transit/VIVA for their forked version of OBA - see [their apps here](http://www.yorkregiontransit.com/en/ridingwithus/apps.asp), and the [code here](https://github.com/tmsyrt/OneBusAway-YORK-Android).  York forked the code, but in this example I've shown how alternate presentation styles can be created in the main OBA code base, and configured for each build flavor.  So, if another brand of OBA wants to take advantage of a different style created by another agency, its as simple as adding a configuration line in `build.gradle`.

Here's a screenshot of how arrival info style B looks in the Agency Y flavor example app:

oba-whitelabel-arrivalstyleb-small

The original OBA and Agency X brands still have the original OBA arrival info presentation.

**Comment by barbeau** [comment]
I'm also seeing this error when hitting the "Load more arrivals" button several times in Agency Y (but I'm not sure its related to Agency Y differences):

~~~
java.lang.NullPointerException: Attempt to read from field 'org.onebusaway.android.io.elements.ObaReferencesElement org.onebusaway.android.io.request.ObaArrivalInfoResponse$Data.references' on a null object reference
            at org.onebusaway.android.io.request.ObaArrivalInfoResponse$Data.access$400(ObaArrivalInfoResponse.java:53)
            at org.onebusaway.android.io.request.ObaArrivalInfoResponse.getStop(ObaArrivalInfoResponse.java:72)
            at org.onebusaway.android.ui.HomeActivity.onArrivalTimesUpdated(HomeActivity.java:652)
            at org.onebusaway.android.ui.ArrivalsListFragment.onLoadFinished(ArrivalsListFragment.java:468)
            at org.onebusaway.android.ui.ArrivalsListFragment.onLoadFinished(ArrivalsListFragment.java:77)
            at android.support.v4.app.LoaderManagerImpl$LoaderInfo.callOnLoadFinished(LoaderManager.java:427)
            at android.support.v4.app.LoaderManagerImpl$LoaderInfo.onLoadComplete(LoaderManager.java:395)
            at android.support.v4.content.Loader.deliverResult(Loader.java:104)
            at org.onebusaway.android.ui.ArrivalsListLoader.deliverResult(ArrivalsListLoader.java:59)
            at org.onebusaway.android.ui.ArrivalsListLoader.deliverResult(ArrivalsListLoader.java:27)
            at android.support.v4.content.AsyncTaskLoader.dispatchOnLoadComplete(AsyncTaskLoader.java:223)
            at android.support.v4.content.AsyncTaskLoader$LoadTask.onPostExecute(AsyncTaskLoader.java:61)
            at android.support.v4.content.ModernAsyncTask.finish(ModernAsyncTask.java:461)
            at android.support.v4.content.ModernAsyncTask.access$500(ModernAsyncTask.java:47)
            at android.support.v4.content.ModernAsyncTask$InternalHandler.handleMessage(ModernAsyncTask.java:474)
            at android.os.Handler.dispatchMessage(Handler.java:102)
            at android.os.Looper.loop(Looper.java:135)
            at android.app.ActivityThread.main(ActivityThread.java:5431)
            at java.lang.reflect.Method.invoke(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:372)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:913)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:706)
~~~

**Comment by barbeau** [comment]
@cagryInside This should be ready for review when you have a chance.  See the README and [REBRANDING.md](https://github.com/OneBusAway/onebusaway-android/blob/branding/REBRANDING.md) file for more details on building the variants.  I'd like you to run the 3 brand variants w/ the Google platform (obaGoogleDebug, agencyXGoogleDebug, and agencyYGoogleDebug) on your Google devices just to make sure everything appears to be working properly.  I tested on Fire Phone with the Amazon platform variants and everything seems to work fine.  Google variants seem to work fine on LG G3 and G4 with Lollipop, but additional testing on your devices would be good (including the EVO 3D, for an older API level).

**Comment by anubiann00b** [comment]
Wow this looks downright beautiful. Can't wait for the merge. Any idea when that'll happen?

I was planning to PR with a swipe down to update thing on the bus arrival times list. Will that feature be in this? It'd probably pointless to add that if this gets merged anyway, no?

**Comment by barbeau** [comment]
@anubiann00b Thanks! :)  This PR should be merged back to `develop` branch shortly - that's where the majority of the new work, including the UI redesign, is taking place.  So, any improvements should be focused there.  I'm hoping to do a push over the next few weeks to finish up the UI redesign (see #177 for remaining items).  As soon as the `develop` branch is feature-equivalent to `master`, it will be merged to `master`.

As far as pull to refresh - this was originally on my list of things to implement in `develop`, but I'm unsure how that will interact with the bottom sliding panel, so it dropped to a lower priority.  I plan to bump the version of the sliding panel library in develop as soon as this PR is merged, and see what the current behavior is for listview scolling within the sliding panel.  I think pull to refresh would work in the `ArrivalsListActivity`, although note that in the new UI `ArrivalsListActivity` is only reachable via shorts from the home screen (in the core UI flow, `ArrivalsListFragment` is shown in the bottom sliding panel from within `HomeActivity`).

In short, please feel free to try implementing pull to refresh for arrival times (in sliding panel and/or `ArrivalsListActivity`), but please use the `develop` branch.  I'd also suggest waiting until this PR is merged, and until I bump the sliding panel library version.  I'll ping you again here once these are complete.

**Comment by anubiann00b** [comment]
Sounds good! It should be fully functional and usable right now, correct?

**Comment by barbeau** [comment]
Cool!  Yes, `develop` branch is stable and functional.

**Comment by cagryInside** [comment]
@barbeau I have tested 3 build flavors with Nexus 6 (v5.1) and HTC evo 3D (v4.0.3). All three flavors works fine with both devices.

Also, I really like setting the arrival info style from gradle settings:
```
 buildConfigField "int", "ARRIVAL_INFO_STYLE", "1"  // Defined in BuildFlavorConstants
```
I have also tested this one with all other flavors and it mainly works fine. However, I found a minor problem with ```obaGoogleDebug``` flavor:

When I use ```"ARRIVAL_INFO_STYLE", "1"```, and click on ```show route information``` button, the new UI comes with navigation drawer opened.

First             |  Second
:-------------------------:|:-------------------------:
device-2015-06-11-140546 | device-2015-06-11-140647

I will be testing further.

**Comment by barbeau** [comment]
Thanks!  The open nav drawer is likely from initial install.  It will open
automatically until you interact with it to make sure the user knows that
it exists.  This was default behavior of the Google IO app I borrowed it
from, and I believe recommended in style guides too.

**Comment by barbeau** [comment]
@cagryInside did you have more testing you wanted to perform before we merge this?  Everything seems fine to me.  Nav drawer behavior is expected at the moment, and sometimes we can change in develop branch if needed.

**Comment by cagryInside** [comment]
@barbeau I looked again this morning, everything works fine.

**Comment by barbeau** [comment]
Thanks @cagryInside!

**Comment by barbeau** [comment]
@anubiann00b Ok, you'll all set to start working on pull to refresh on the current `develop` branch.  As of commit https://github.com/OneBusAway/onebusaway-android/commit/944f16f48723dbc1b369f36b90624ffbafebeb0e, it includes both the rebranding feature and an update to the latest AndroidSlidingUpPanel library v3.0.0.  I would suggest opening a new ticket for pull to refresh indicating that you're working on this, and tracking your progress via a WIP pull request.

**Comment by anubiann00b** [comment]
With the current sliding panel, I think the change I was proposing would be redundant. I was thinking of the current state of master, where the time list is completely seperate from the map. Maybe not though...

**Comment by barbeau** [comment]
Pull to refresh would still be useful when opening the arrivals list from a shortcut on the home screen, since in that case its not presented in a sliding panel, but instead very similarly to the current production version.

Current sliding panel behavior is to scroll the list when the sliding panel is open, but it does not collapse when pulling down and reaching the top of the list.  This current behavior of the sliding panel is a little strange with lists (sometimes on high velocity scrolls it does collapse), and its something I plan to look more into - given the current library we're using, I'm not sure if it can be changed.  If it can't be changed, then I think pull to refresh there may be useful.
